src/Lib.P2P/SecureCommunication/Secio1Stream.cs

```diff
@@ -0,0 +1,270 @@
+                if (_inBlock != null && _inBlockOffset < _inBlock.Length)
+                {
+                    var n = Math.Min(_inBlock.Length - _inBlockOffset, count);
+                    Array.Copy(_inBlock, _inBlockOffset, buffer, offset, n);
+                    total += n;
+                    count -= n;
+                    offset += n;
+                    _inBlockOffset += n;
+                }
+
+                // Otherwise, wait for a new block of data.
+                else
+                {
+                    _inBlock = await ReadPacketAsync(cancellationToken);
+                    _inBlockOffset = 0;
+                }
+
+            return total;
+        }
+
+        /// <summary>
+        ///   Read an encrypted and signed packet.
+        /// </summary>
+        /// <returns>
+        ///   The plain text as an array of bytes.
+        /// </returns>
+        /// <remarks>
+        ///   A packet consists of a [uint32 length of packet | encrypted body | hmac signature of encrypted body].
+        /// </remarks>
+        private async Task<byte[]> ReadPacketAsync(CancellationToken cancel)
+        {
+            var lengthBuffer = await ReadPacketBytesAsync(4, cancel).ConfigureAwait(false);
+            var length =
+                (lengthBuffer[0] << 24) |
+                (lengthBuffer[1] << 16) |
+                (lengthBuffer[2] << 8) |
+                lengthBuffer[3];
+            if (length <= _outHmac.GetMacSize())
+                throw new InvalidDataException($"Invalid secio packet length of {length}.");
+
+            var encryptedData = await ReadPacketBytesAsync(length - _outHmac.GetMacSize(), cancel).ConfigureAwait(false);
+            var signature = await ReadPacketBytesAsync(_outHmac.GetMacSize(), cancel).ConfigureAwait(false);
+
+            var hmac = _outHmac;
+            var mac = new byte[hmac.GetMacSize()];
+            hmac.Reset();
+            hmac.BlockUpdate(encryptedData, 0, encryptedData.Length);
+            hmac.DoFinal(mac, 0);
+            if (!signature.SequenceEqual(mac))
+                throw new InvalidDataException("HMac error");
+
+            // Decrypt the data in-place.
+            _decrypt.ProcessBytes(encryptedData, 0, encryptedData.Length, encryptedData, 0);
+            return encryptedData;
+        }
+
+        private async Task<byte[]> ReadPacketBytesAsync(int count, CancellationToken cancel)
+        {
+            var buffer = new byte[count];
+            await _stream.ReadExactAsync(buffer, 0, count, cancel).ConfigureAwait(false);
+            return buffer;
+        }
+
+        /// <inheritdoc />
+        public override void Flush()
+        {
+#pragma warning disable VSTHRD002 
+            FlushAsync().GetAwaiter().GetResult();
+#pragma warning restore VSTHRD002 
+        }
+
+        /// <inheritdoc />
+        public override async Task FlushAsync(CancellationToken cancel)
+        {
+            if (_outStream.Length == 0)
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Add curly braces around the nested statement(s) in this 'if' block.](https://app.codacy.com/manual/catalyst-network/Catalyst.Framework/pullRequest?prid=4832783)